src/types.cpp

```diff
@@ -110,85 +110,76 @@ void DataBuf::reset() {
   }
   return getUShort(&pData_[offset], byteOrder);
 }
 
 void Exiv2::DataBuf::write_uint16(size_t offset, uint16_t x, ByteOrder byteOrder) {
   if (pData_.size() < 2 || offset > (pData_.size() - 2)) {
-    throw std::overflow_error("Overflow in Exiv2::DataBuf::write_uint16");
+    throw std::out_of_range("Overflow in Exiv2::DataBuf::write_uint16");
   }
   us2Data(&pData_[offset], x, byteOrder);
 }
 
 uint32_t Exiv2::DataBuf::read_uint32(size_t offset, ByteOrder byteOrder) const {
   if (pData_.size() < 4 || offset > (pData_.size() - 4)) {
-    throw std::overflow_error("Overflow in Exiv2::DataBuf::read_uint32");
+    throw std::out_of_range("Overflow in Exiv2::DataBuf::read_uint32");
   }
   return getULong(&pData_[offset], byteOrder);
 }
 
 void Exiv2::DataBuf::write_uint32(size_t offset, uint32_t x, ByteOrder byteOrder) {
   if (pData_.size() < 4 || offset > (pData_.size() - 4)) {
-    throw std::overflow_error("Overflow in Exiv2::DataBuf::write_uint32");
+    throw std::out_of_range("Overflow in Exiv2::DataBuf::write_uint32");
   }
   ul2Data(&pData_[offset], x, byteOrder);
 }
 
 uint64_t Exiv2::DataBuf::read_uint64(size_t offset, ByteOrder byteOrder) const {
   if (pData_.size() < 8 || offset > (pData_.size() - 8)) {
-    throw std::overflow_error("Overflow in Exiv2::DataBuf::read_uint64");
+    throw std::out_of_range("Overflow in Exiv2::DataBuf::read_uint64");
   }
   return getULongLong(&pData_[offset], byteOrder);
 }
 
 void Exiv2::DataBuf::write_uint64(size_t offset, uint64_t x, ByteOrder byteOrder) {
   if (pData_.size() < 8 || offset > (pData_.size() - 8)) {
-    throw std::overflow_error("Overflow in Exiv2::DataBuf::write_uint64");
+    throw std::out_of_range("Overflow in Exiv2::DataBuf::write_uint64");
   }
   ull2Data(&pData_[offset], x, byteOrder);
 }
 
 int Exiv2::DataBuf::cmpBytes(size_t offset, const void* buf, size_t bufsize) const {
   if (pData_.size() < bufsize || offset > pData_.size() - bufsize) {
-    throw std::overflow_error("Overflow in Exiv2::DataBuf::cmpBytes");
+    throw std::out_of_range("Overflow in Exiv2::DataBuf::cmpBytes");
   }
   return memcmp(&pData_[offset], buf, bufsize);
 }
 
 byte* Exiv2::DataBuf::data(size_t offset) {
-  /// \todo this first check should be for <= offset
-  if (pData_.size() < offset) {
-    throw std::overflow_error("Overflow in Exiv2::DataBuf::c_data");
-  }
-  if (pData_.empty() || pData_.size() == offset) {
-    return nullptr;
-  }
-  return &pData_[offset];
+  return const_cast<byte*>(c_data(offset));
 }
 
 const byte* Exiv2::DataBuf::c_data(size_t offset) const {
-  /// \todo this first check should be for <= offset
-  if (pData_.size() < offset) {
-    throw std::overflow_error("Overflow in Exiv2::DataBuf::c_data");
-  }
-  if (pData_.empty() || pData_.size() == offset) {
+  if (pData_.empty()) {
     return nullptr;
+  } else if (offset >= pData_.size()) {
+    throw std::out_of_range("Overflow in Exiv2::DataBuf::c_data");
```

Comment:
I'm not sure that this is a good change. It has created a lot of special cases in the code. For example, this code:

```
out.write(iccProfile_.c_str(), iccProfile_.size());
```

had to be replaced by this:

```
out.write(iccProfile_.empty() ? nullptr : iccProfile_.c_str(), iccProfile_.size());
```

If we want to do accurate overflow checking here, then maybe we could instead add a method that does something like this:

```
const byte* Exiv2::DataBuf::safe_c_data(size_t offset, size_t len) const {
   if (Safe::add(offset, len) > pData_.size()) {
    throw std::out_of_range("Overflow in Exiv2::DataBuf::safe_c_data");
  }
  return &pData_[offset];
}

Reply:
I also had the same thought as you when I was implementing some of the changes.

In fact, afterwards (commit 6964f5f9f2f83db92d94287c0f0d380db7006585) I modified slightly the `c_data()` method to return `nullptr` when DataBuf is empty. I can do the same for `c_str()`. 

In that particular case that you mentioned, I think it makes sense to leave the code as it was. There were some other cases in which it makes to prevent additional work in case DataBuf is empty.

I will create a new commit with these changes.

Reply:
Done in 2b74cc885d4825aa3927fea9125b8250f3a51f94

Reply:
I would prefer this:

```suggestion
  } else if (offset > pData_.size()) {
    throw std::out_of_range("Overflow in Exiv2::DataBuf::c_data");
```

Just because I think `offset == pData_.size()` is fine if you're going to read zero bytes. That's an edge case that can happen quite easily in correct code.

Reply:
If I make the change you proposed here, the following test which I added in this branch fails:

```cpp
TEST(DataBuf, tryingToAccessTooFarElementThrows) {
  const std::array<byte, 4> data{'h', 'o', 'l', 'a'};
  DataBuf instance(data.data(), data.size());
  ASSERT_THROW([[maybe_unused]] auto d = instance.data(4), std::out_of_range);
  ASSERT_THROW([[maybe_unused]] auto d = instance.c_data(4), std::out_of_range);
}
```

You could be trying to access an element past the last one in the `DataBuf` instance, and there would not be a way to know if the returned address is `nullptr` or some random value. Note that the interface of this function actually differs to `std::vector::data()` or `std::string::c_str()` in which we are providing an offset parameter. I think we should check if the offset provided makes sense or not.

You are right in pointing out that in most of cases when we use DataBuf like this, we end up trying to read 0 bytes, and therefore we do not have a problem with this. But I think it would not be difficult to mix a wrong usage of `DataBuf.c_data(offset)` and then try to read a number of bytes given by other precomputed variable. 

In the last months I have been reducing the LoCs of DataBuf considerably and I have the feeling that we could actually drop it and replace it by usages of `std::vector` or `std::array` without much trouble. At the end `DataBuf` is just a wrapper around `std::vector` with some helper functions to read/write integers of different sizes. 

I'll keep the code as it is since this is anyways a minor issue.